clippy_lints/src/needless_continue.rs

```diff
@@ -0,0 +1,467 @@
+                func(e, cond, if_block, else_expr);
+            }
+        },
+        _ => { },
+    }
+}
+
+/// A type to distinguish between the two distinct cases this lint handles.
+enum LintType {
+    ContinueInsideElseBlock,
+    ContinueInsideThenBlock,
+}
+
+/// Data we pass around for construction of help messages.
+struct LintData<'a> {
+    if_expr:     &'a ast::Expr,    // The `if` expr encountered in the above loop.
+    if_cond:     &'a ast::Expr,    // The condition expression for the above `if`.
+    if_block:    &'a ast::Block,   // The `then` block of the `if` statement.
+    else_expr:   &'a ast::Expr,    /* The `else` block of the `if` statement.
+                                      Note that we only work with `if` exprs that
+                                      have an `else` branch. */
+    stmt_idx:    usize,            /* The 0-based index of the `if` statement in
+                                      the containing loop block. */
+    block_stmts: &'a [ast::Stmt],  // The statements of the loop block.
+}
+
+const MSG_REDUNDANT_ELSE_BLOCK: &'static str = "This else block is redundant.\n";
+
+const MSG_ELSE_BLOCK_NOT_NEEDED: &'static str = "There is no need for an explicit `else` block for this `if` expression\n";
+
+const DROP_ELSE_BLOCK_AND_MERGE_MSG: &'static str =
+    "Consider dropping the else clause and merging the code that follows (in the loop) with the if block, like so:\n";
+
+const DROP_ELSE_BLOCK_MSG: &'static str =
+    "Consider dropping the else clause, and moving out the code in the else block, like so:\n";
+
+
+fn emit_warning<'a>(ctx: &EarlyContext,
+                    data: &'a LintData,
+                    header: &str,
+                    typ: LintType) {
+
+    // snip    is the whole *help* message that appears after the warning.
+    // message is the warning message.
+    // expr    is the expression which the lint warning message refers to.
+    let (snip, message, expr) = match typ {
+        ContinueInsideElseBlock => {
+            (suggestion_snippet_for_continue_inside_else(ctx, data, header),
+             MSG_REDUNDANT_ELSE_BLOCK,
+             data.else_expr)
+        },
+        ContinueInsideThenBlock => {
+            (suggestion_snippet_for_continue_inside_if(ctx, data, header),
+             MSG_ELSE_BLOCK_NOT_NEEDED,
+             data.if_expr)
+        }
+    };
+    span_help_and_lint(ctx, NEEDLESS_CONTINUE, expr.span, message, &snip);
+}
+
+fn suggestion_snippet_for_continue_inside_if<'a>(ctx: &EarlyContext,
+                                                data: &'a LintData,
+                                                header: &str) -> String {
+    let cond_code = &snippet(ctx, data.if_cond.span, "..").into_owned();
+
+    let if_code   = format!("if {} {{\n    continue;\n}}\n", cond_code);
+                                   /*  ^^^^--- Four spaces of indentation. */
+    // region B
```

Comment:
There should be a comment somewhere explaining what those regions are for those you haven't read or don't remember the first post of this PR :smile: